zebra/zebra_vxlan.c

```diff
@@ -9062,7 +9062,9 @@ void zebra_vxlan_advertise_all_vni(ZAPI_HANDLER_ARGS)
 		return;
 
 	zvrf->advertise_all_vni = advertise;
-	if (is_evpn_enabled()) {
+	if (zvrf->advertise_all_vni) {
```

Comment:
Old code checks if EVPN is enabled or not. If not enabled, it does the some cleanup. With the new code when advertise_all_vni is not set on this VRF, then  we are assuming that it is enabled on Default VRF. 

What is advertise_all_vni flag in default VRF is not set when we are setting the zvrf_default->evpn_vrf_id to default VRF ID ? I am not sure if this is right.

Reply:
> when advertise_all_vni is not set on this VRF, then we are assuming that it is enabled on Default VRF.

I don't get how we are assuming this. The cleanup is done on the VRF on which this function is applied, not on the default one (which is why the reverting of `evpn_vrf_id` to default is at the end of the `else` branch).

Moreover, thanks to the check L9061, the function will early exit on any vrf that is not the evpn one, no matter if any are configured to do so or not.

> What is advertise_all_vni flag in default VRF is not set when we are setting the zvrf_default->evpn_vrf_id to default VRF ID ? I am not sure if this is right.

The flag cannot be set, as safety checks in the vty handler prevent the flag being set on more than one vrf. And we are setting the evpn_vrf_id to default since it has to have a value, and it is the only one that we are sure will always point to an existing vrf.

Say we don't set it to the default vrf, and then the vrf gets deleted. We now have a "dangling pointer", so to speak. By setting it to default vrf, we know we can always call `zebra_vrf_lookup_by_id` on the `evpn_vrf_id` and will always receive a valid pointer.

Reply:
Now I understand your change on why you are setting this to VRF_DEFAULT. However, what if the EVPN is actually only in VRF A, and now disabled on VRF A? It was never even enabled on default VRF at all. Do you think setting this to default_VRF is the right thing to do ?

Reply:
(I will probably get flame for writing this, but oh well...) I have tu put _something_ in this variable. I cannot leave it to "whatever was the evpn vrf" because it could end up being "dangling". I could put an invalid VRF value in there, but which one? I didn't find any, and only two values are special:
- 0 which is the default vrf when the backend is vrf, and a valid ID when the backend is netns
- UINT32_MAX when the backend is netns, and a valid ID when the backend is vrf (well in fact good luck reaching this as a vrf ifindex in linux in practice, but nothing forbids it)
And negative values are out of the question since vrf_id_t is actually uint32_t.

So yes, for the time being I think setting this to DEFAULT_VRF is right (but not necessarily _the_ right nor _the best_ thing to do), especially since "evpn enabled" <=> default_vrf->evpn_vrf->advertise_all_vni == true.

Maybe renaming the variable to `evpn_vrf_id_if_evpn_is_enabled` would make more sense? (overly exaggerating, but this is the meaning of this variable). Or adding a second one, which would be a flag to tell wether evpn is enabled, and explicitly telling to consider the contents of `evpn_vrf_id` if the flag is true.

In the end, all I want for this variable is to _always_ hold a valid VRF ID.

Reply:
Oh and I just remembered that [@AnuradhaKaruppiah's comment](https://github.com/FRRouting/frr/pull/3923#issuecomment-474387098) explicitly confirms that, in absence of an EVPN VRF, we should get the default one. There, storing the default one there makes sense